design/dpop.md

```diff
@@ -0,0 +1,308 @@
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.endpoints.TokenEndpoint
+  * processGrantRequest()
+
+#### Introspection Endpoint
+
+For DPoP-bound tokens, token introspection endpoint must return an additional `cnf` claim with the `jkt` member containing JWK thumbprint (a SHA-256 hash).
+
+_([6.2. JWK Thumbprint Confirmation Method in Token Introspection][8])_
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.AccessTokenIntrospectionProvider
+  * introspect()
+
+#### Metadata Endpoint
+
+With DPoP enabled, a new parameter, `dpop_signing_alg_values_supported`, must be included into the authorization server metadata.
+
+_([5.1. Authorization Server Metadata][10])_
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.OIDCWellKnownProvider
+	* getConfig()
+* org.keycloak.protocol.oidc.representations.OIDCConfigurationRepresentation
+
+#### UserInfo Endpoint
+
+The spec doesn’t directly mention the UserInfo endpoint, but it clearly falls into the category of bearer token protected resources:
+
+> The UserInfo Endpoint is an OAuth 2.0 Protected Resource that returns Claims about the authenticated End-User. To obtain the requested Claims about the End-User, the Client makes a request to the UserInfo Endpoint using an Access Token obtained through OpenID Connect Authentication.
+
+_(OpenID Connect Core 1.0,_ [_5.3. UserInfo Endpoint_][11]_)_
+
+Therefore, UserInfo endpoint must also be made DPoP-aware.
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.endpoints.UserInfoEndpoint
+	* issueUserInfo()
+
+#### Logout Endpoint
+
+The `POST` variant of logout invocation uses refresh token, which could be DPoP-bound. Therefore, Logout endpoint must also be made DPoP-aware.
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.endpoints.LogoutEndpoint
+	* logoutToken()
+
+#### Other Endpoints
+
+The following endpoints in Keycloak are public:
+
+* Server Discovery
+* Server JWK set
+* Authorization
+
+The following endpoints don’t use bearer tokens, but rather client credentials:
+
+* Token Introspection
+* Token Revocation
+
+The following endpoints use cookie authentication:
+
+* Check Session iframe
+
+Therefore, no other endpoints should be affected.
+
+#### Client Registration
+Client registration should support DPoP-related parameters inside client metadata.
+
+Classes/methods affected:
+
+* org.keycloak.services.clientregistration.oidc.DescriptionConverter
+* org.keycloak.representations.oidc.OIDCClientRepresentation
+
+### Models
+
+#### Access Token
+
+New property, `jkt`, should be recognized as a member of the `cnf` claim.
+
+Classes/methods affected:
+
+* org.keycloak.representations.AccessToken.CertConf
+	* rename to Conf or Confirmation; introduce `keyThumbprint` property
+
+#### Client
+
+There are the following DPoP-related client settings:
+
+* DPoP mode (enabled/optional/disabled)
```

Comment:
To simplify the implementation, it is only needed to add one client settings:
- dpop_bound_access_tokens (true/false)
dpop_bound_access_tokens is a client metadata defined in the DPoP specification.
The boolean value is easy to handle compared with ternary (enabled/optional/disabled).

As for DPoP proof lifetime and clock skew, it might be better to use default values for all clients. It might be better not to add new client settings because keycloak has already a lot of such settings and these settings are difficult to be managed.

If we add client settings at this time and we add some method (e.g., client policies) to remove necessities of the settings afterwards,  we cannnot delete these settings because we need to keep backward compatibility.

Therefore, at this time, it might be preferable to use default values only for DPoP proof lifetime and clock skew, and add some mechanism to change these values per client or set of clients without adding client settings afterwards.